core/dbt/clients/_jinja_blocks.py

```diff
@@ -417,156 +246,122 @@ def next_block(self, block_handler, include_toplevel=False):
-            elif matchgroups.get('open'):
-                self._parenthesis_stack.append(True)
-            elif matchgroups.get('close'):
-                self._parenthesis_stack.pop()
-            else:
-                raise dbt.exceptions.InternalException(
-                    'unhandled regex in _process_macro_default_arg(), no match'
-                    ': {}'.format(matchgroups)
-                )
 
-    def _process_macro_args(self):
-        """Macro args are pretty tricky! Arg names themselves are simple, but
-        you can set arbitrary default values, including doing stuff like:
-        {% macro my_macro(arg="x" + ("}% {# {% endmacro %}" * 2)) %}
+class BlockIterator(object):
+    def __init__(self, data):
+        self.tag_parser = TagIterator(data)
+        self.current = None
+        self.stack = []
+        self.last_position = 0
+
+    @property
+    def current_end(self):
+        if self.current is None:
+            return 0
+        else:
+            return self.current.end
 
-        Which makes you a jerk, but is valid jinja.
-        """
-        # we are currently after the first parenthesis (+ any whitespace) after
-        # the macro args started. You can either have the close paren, or a
-        # name.
-        while self._parenthesis_stack:
-            match = self._expect_match('macro arguments',
-                                       MACRO_ARGS_END_PATTERN,
-                                       MACRO_ARG_PATTERN)
-            self.advance(match.end())
-            matchgroups = match.groupdict()
-            if matchgroups.get('macro_end') is not None:
-                self._parenthesis_stack.pop()
-            # we got an argument. let's see what it has
-            elif matchgroups.get('value') is not None:
-                # we have to process a single macro argument. This mutates
-                # the parenthesis stack! If it finds a comma, it will continue
-                # the loop.
-                self._process_macro_default_arg()
-            elif matchgroups.get('more_args') is not None:
-                continue
-            else:
-                raise dbt.exceptions.InternalException(
-                    'unhandled regex in _process_macro_args(), no match: {}'
-                    .format(matchgroups)
+    @property
+    def data(self):
+        return self.tag_parser.data
+
+    def is_current_end(self, tag):
+        return (
+            tag.block_type_name.startswith('end') and
+            self.current is not None and
+            tag.block_type_name[3:] == self.current.block_type_name
+        )
+
+    def find_blocks(self, allowed=None, collect_raw_data=True):
+        """Find all top-level blocks in the data."""
+        if allowed is None:
+            allowed = {'snapshot', 'macro', 'materialization', 'docs'}
+
+        for tag in self.tag_parser.find_tags():
+            if tag.block_type_name in _CONTROL_FLOW_TAGS:
+                self.stack.append(tag.block_type_name)
```

Comment:
for some reason when we discussed this before i hadn't considered using a stack to track descent (even though this is like... the way to do this thing?)

i'm really happy with how this code turned out, and i agree that this is going to be much easier to maintain.